hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java

```diff
@@ -261,6 +261,12 @@ public void renameTable(TableIdentifier from, TableIdentifier originalTo) {
     } catch (InterruptedException e) {
       Thread.currentThread().interrupt();
       throw new RuntimeException("Interrupted in call to rename", e);
+    } catch (RuntimeException e) {
```

Comment:
Can you paste the complete callstack? I am wondering why it is not coming as `AlreadyExistsException` in line 254

Reply:
`java.lang.AssertionError: 
Expecting actual throwable to be an instance of:
  org.apache.iceberg.exceptions.AlreadyExistsException
but was:
  java.lang.RuntimeException: InvalidOperationException(message:new table newdb.table_renamed already exists)
	at org.apache.iceberg.relocated.com.google.common.base.Throwables.propagate(Throwables.java:231)
	at org.apache.iceberg.common.DynMethods$UnboundMethod.invoke(DynMethods.java:75)
	at org.apache.iceberg.hive.MetastoreUtil.alterTable(MetastoreUtil.java:75)`

Reply:
Taking a quick look, I do not like this change.

Why is this `RuntimeException`?

This code tries to capture this case:
```
    } catch (AlreadyExistsException e) {
      throw new org.apache.iceberg.exceptions.AlreadyExistsException(
          "Table already exists: %s", to);
    }
```

Do we have checks for the other cases as well?
My first guess would be that `MetastoreUtil.alterTable` messes up the Hive exceptions, and maybe all of them are off?

Reply:
Also maybe the exceptions thrown are different with different Hive versions?

Reply:
I would be surprised if the HMS API would throw a RuntimeException.
Some Iceberg code might convert it to a RuntimeException, but the HMS API usually throws MetastoreExceptions

Reply:
<img width="1728" alt="Screenshot 2023-11-10 at 4 03 25 PM" src="https://github.com/apache/iceberg/assets/4146188/d0575ade-1512-4d1b-853f-9a26e35fc343">
This is how the stacktrace elements are looking for HMS alter table ops.

Reply:
So we know that the HMS throws an `InvalidOperationException`, but I guess that the `MetastoreUtil.alterTable` wraps the exception to a `RuntimeException`.

I guess that this is also true for the previous exceptions thrown by HMS:
```
    } catch (NoSuchObjectException e) {
      throw new NoSuchTableException("Table does not exist: %s", from);
    } catch (AlreadyExistsException e) {
      throw new org.apache.iceberg.exceptions.AlreadyExistsException(
          "Table already exists: %s", to);
    } catch (TException e) {
      throw new RuntimeException("Failed to rename " + from + " to " + to, e);
    }
```

Do we have tests for those?

------

What I trying to understand if any of the HMS errors are handled correctly, or we only have to fix specific case

Reply:
Yes, It seems `AlterTable` is wrapping exception into `RuntimeException` [here](https://github.com/apache/iceberg/blob/bfe1d03c75b4a8bd2c15c925f39cdec28c312489/common/src/main/java/org/apache/iceberg/common/DynMethods.java#L70) . 
We do have tests for other exceptions except `AlreadyExistsException`. Also I am not sure `AlreadyExistsException` exception is being thrown for `renameTable`.  Although `NoSuchObjectException` and `TException` is being thrown for retrieving the table data.

Reply:
Created a cleanup [PR](https://github.com/apache/iceberg/pull/9143) in terms of error propagation.

Reply:
I would wait a bit if some more knowledgable around #9143 would review that PR.

> Yes, It seems AlterTable is wrapping exception into RuntimeException [here](https://github.com/apache/iceberg/blob/bfe1d03c75b4a8bd2c15c925f39cdec28c312489/common/src/main/java/org/apache/iceberg/common/DynMethods.java#L70) .
We do have tests for other exceptions except AlreadyExistsException. Also I am not sure AlreadyExistsException exception is being thrown for renameTable. Although NoSuchObjectException and TException is being thrown for retrieving the table data.

If `NoSuchObjectException` and `TException` is thrown - do we handle it correctly? Does it wrap into `RuntimeException`? Do we unwrap it before checking?

Reply:
Yes, [NoSuchObjectException](https://github.com/apache/iceberg/blob/f246614ed9d98f84eae13ff57768d377a95d3ced/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java#L251) and [TException](https://github.com/apache/iceberg/blob/f246614ed9d98f84eae13ff57768d377a95d3ced/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java#L258)  both are checked exceptions and being handled explicitly.

Reply:
@pvary @nastra is it a good idea for now to override the corresponding test cases at `TestHiveCatalog`. and remove this specific `RuntimeException` handling ? 
Currently there is only one test cases which is failing [testRenameTableDestinationTableAlreadyExists](https://github.com/apache/iceberg/blob/7240752e18278b3454be141c2e4b121079004d8b/core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java#L779). 
It is failing [here](https://github.com/apache/iceberg/blob/7240752e18278b3454be141c2e4b121079004d8b/core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java#L803). 

With a backlog issue we can fix this `RuntimeException` issue. 
WDYT ?

Reply:
I'm ok ignoring `testRenameTableDestinationTableAlreadyExists()` for `TestHiveCatalog` for now and revert the changes here and then open an issue to follow-up on improving this behavior.

Reply:
Created an [issue](https://github.com/apache/iceberg/issues/9289) to track the same.  @nastra , Please assign this issue to me. Meantime, I will look into this.